dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AnalysisController.java

```diff
@@ -0,0 +1,695 @@
+    {
+        for ( DataElementOperand operand : expressionService
+            .getOperandsInExpression( validationRule.getLeftSide().getExpression() ) )
+        {
+            DataValue dataValue = dataValueService
+                .getDataValue( operand.getDataElement(), period, organisationUnit, operand.getCategoryOptionCombo() );
+
+            String value = dataValue != null ? dataValue.getValue() : null;
+
+            validationRuleExpressionDetails.addLeftSideDetail( operand.getName(), value );
+        }
+    }
+
+    private void processRightSideDetails( ValidationRuleExpressionDetails validationRuleExpressionDetails,
+        ValidationRule validationRule, OrganisationUnit organisationUnit, Period period )
+    {
+        for ( DataElementOperand operand : expressionService
+            .getOperandsInExpression( validationRule.getRightSide().getExpression() ) )
+        {
+            DataValue dataValue = dataValueService
+                .getDataValue( operand.getDataElement(), period, organisationUnit, operand.getCategoryOptionCombo() );
+
+            String value = dataValue != null ? dataValue.getValue() : null;
+
+            validationRuleExpressionDetails.addRightSideDetail( operand.getName(), value );
+        }
+    }
+
+    private Grid generateAnalysisReportGridFromResults( List<DeflatedDataValue> results, OrganisationUnit orgUnit )
+    {
+        Grid grid = new ListGrid();
+        if ( results != null )
+        {
+            I18nFormat format = i18nManager.getI18nFormat();
+            I18n i18n = i18nManager.getI18n();
+
+            grid.setTitle( i18n.getString( "data_analysis_report" ) );
+
+            if ( orgUnit != null )
+            {
+                grid.setSubtitle( orgUnit.getName() );
+            }
+
+            grid.addHeader( new GridHeader( i18n.getString( "dataelement" ), false, true ) );
+            grid.addHeader( new GridHeader( i18n.getString( "source" ), false, true ) );
+            grid.addHeader( new GridHeader( i18n.getString( "period" ), false, true ) );
+            grid.addHeader( new GridHeader( i18n.getString( "min" ), false, false ) );
+            grid.addHeader( new GridHeader( i18n.getString( "value" ), false, false ) );
+            grid.addHeader( new GridHeader( i18n.getString( "max" ), false, false ) );
+
+            for ( DeflatedDataValue dataValue : results )
+            {
+                Period period = dataValue.getPeriod();
+
+                grid.addRow();
+                grid.addValue( dataValue.getDataElementName() );
+                grid.addValue( dataValue.getSourceName() );
+                grid.addValue( format.formatPeriod( period ) );
+                grid.addValue( dataValue.getMin() );
+                grid.addValue( dataValue.getValue() );
+                grid.addValue( dataValue.getMax() );
+            }
+        }
+
+        return grid;
+    }
+
+    private Grid generateValidationRulesReportGridFromResults( List<ValidationResult> results,
```

Comment:
This method should be pushed down to the service layer (service implementation class) and out of the web api controller layer.

Reply:
This method is only used to "build" the grid we want to export for Validation Rules. We can put this method inside ValidationRuleService. However isn't this presentational logic, and thus be left in the controller?
In spite of that, there is another similar method named `generateAnalysisReportGridFromResults` used to export Std Dev Outlier Analysis, Min-Max Outlier Analysis, and Follow-Up Analysis reports. In this case, should we create a new service to move this method into, or, do you have another suggestion?

Reply:
OK we can leave it.